src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs

```diff
@@ -772,13 +774,13 @@ private static void AddModule(PEModule module, int referenceIndex, ResolvedRefer
             TCompilation compilation,
             DiagnosticBag diagnostics,
             out ImmutableArray<MetadataReference> references,
-            out IDictionary<(string, string), MetadataReference> boundReferenceDirectives,
+            [NotNull] out IDictionary<(string, string), MetadataReference>? boundReferenceDirectives,
             out ImmutableArray<Location> referenceDirectiveLocations)
         {
-            boundReferenceDirectives = null;
+            boundReferenceDirectives = null!;
```

Comment:
@jcouv this `!` was needed once I marked the parameter as `[NotNull]`. This doesn't line up to my intuition though. My intuition is that `[NotNull]` should ensure the state of the parameter on method exit, not transition states within the method body. Am I thinking wrong here?

Reply:
The current enforcement for `[NotNull]` is that you cannot assign null to it. For `[NotNullWhen]` it was enforced on exit, because we don't have a choice.

I'm happy to adjust if you think we should enforce both upon exit. But the reason I kept `[NotNull]` enforcement like this is to warn closer to the problem.

Reply:
My intuition has been that `[NotNull]` was a promise to the caller. Essentially when this method is complete that value will be non-null. Hence as a method implementer a `null` intermediate state is okay so long as it's non-null at exit.

My worry with the current implementation is that the intermediate states are now dangerous. If I have to use `null!` then I've lost the value of `[NotNull]` because the value is now registered as non-null. Hence I don't get the exit guard that I expected. 

Is there a better pattern I should be using here that I'm not considering?

Reply:
> Is there a better pattern I should be using here that I'm not considering?

I would suggest using a temp variable, and no suppression.

Reply:
Went ahead and took the local approach.